File: backend-go/internal/data/tenant_rag.go
package data

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

const (
	defaultTenantRAGTopK = 6
	vectorMetaApp        = "_csm"
	vectorMetaTable      = "vector_chunks"
)

// TenantRAGChunk is one indexed tenant context document for AI retrieval.
type TenantRAGChunk struct {
	ChunkID     string
	AppID       string
	SourceName  string
	ScopeMask   int
	ScopeTags   string
	Tags        string
	CreatedAtMs int64
	Summary     string
	Structure   string
	Content     string
}

// TenantRAGHit is a retrieval result row.
type TenantRAGHit struct {
	ChunkID     string
	AppID       string
	SourceName  string
	ScopeMask   int
	Tags        string
	Summary     string
	Content     string
	Score       float64
	CreatedAtMs int64
}

// EnsureTenantRAGSchema ensures chromem collections are ready (no SQLite).
func (rm *RecordManager) EnsureTenantRAGSchema() error {
	if rm.vectorStore == nil {
		return fmt.Errorf("vector store unavailable")
	}
	_, err := rm.vectorStore.collection(vectorCollTenantRAG)
	return err
}

// UpsertTenantRAGChunk indexes one tenant RAG chunk in chromem (+ optional Pebble mirror).
func (rm *RecordManager) UpsertTenantRAGChunk(chunk TenantRAGChunk) error {
	if err := rm.EnsureTenantRAGSchema(); err != nil {
		return err
	}
	if chunk.ChunkID == "" || chunk.AppID == "" || strings.TrimSpace(chunk.Content) == "" {
		return nil
	}
	meta := tenantChunkMeta(chunk)
	text := tenantChunkEmbedText(chunk)
	if err := rm.vectorStore.upsertDoc(vectorCollTenantRAG, chunk.ChunkID, meta, text); err != nil {
		return err
	}
	rm.mirrorVectorChunkPebble(chunk)
	return nil
}

// DeleteTenantRAGSource removes all chunks for a source within an app.
func (rm *RecordManager) DeleteTenantRAGSource(appID, sourceName string) error {
	if rm.vectorStore == nil || appID == "" || sourceName == "" {
		return nil
	}
	_ = rm.vectorStore.deleteWhere(vectorCollTenantRAG, map[string]string{
		"app_id":      appID,
		"source_name": sourceName,
	})
	rm.deleteVectorChunksPebbleBySource(appID, sourceName)
	return nil
}

// SearchTenantRAG runs semantic vector search (chromem) with scope mask filter.
// queryText is natural language (not FTS match syntax).
func (rm *RecordManager) SearchTenantRAG(appID, queryText string, scopeMask, limit int) ([]TenantRAGHit, error) {
	if rm.vectorStore == nil {
		return nil, fmt.Errorf("vector store unavailable")
	}
	if appID == "" || strings.TrimSpace(queryText) == "" {
		return nil, nil
	}
	if limit <= 0 {
		limit = defaultTenantRAGTopK
	}
	results, err := rm.vectorStore.query(vectorCollTenantRAG, queryText, map[string]string{"app_id": appID}, limit*3)
	if err != nil {
		return nil, err
	}
	var hits []TenantRAGHit
	for _, r := range results {
		h := chromemHitToTenantRAG(r)
		if !scopeMaskMatches(strconv.Itoa(h.ScopeMask), scopeMask) {
			continue
		}
		hits = append(hits, h)
		if len(hits) >= limit {
			break
		}
	}
	return hits, nil
}

// SearchRecordsVectorForApp semantic search over indexed org/menu tables (replaces records_fts RAG leg).
func (rm *RecordManager) SearchRecordsVectorForApp(appID, queryText string, tableNames []string, limit int) ([]TenantRAGHit, error) {
	if rm.vectorStore == nil || appID == "" || strings.TrimSpace(queryText) == "" {
		return nil, nil
	}
	if limit <= 0 {
		limit = defaultTenantRAGTopK
	}
	if len(tableNames) == 0 {
		tableNames = []string{"csm_roles", "csm_depts", "csm_branches", "index", "sys_autos"}
	}
	tableSet := make(map[string]struct{}, len(tableNames))
	for _, t := range tableNames {
		tableSet[t] = struct{}{}
	}
	results, err := rm.vectorStore.query(vectorCollRecords, queryText, map[string]string{"app_id": appID}, limit*4)
	if err != nil {
		return nil, err
	}
	var hits []TenantRAGHit
	for _, r := range results {
		tableName := r.Metadata["table_name"]
		if _, ok := tableSet[tableName]; !ok {
			continue
		}
		scope := scopeMaskForTable(tableName)
		hits = append(hits, TenantRAGHit{
			ChunkID:     r.ID,
			AppID:       appID,
			SourceName:  tableName,
			ScopeMask:   scope,
			Summary:     r.Metadata["title"],
			Content:     r.Content,
			Score:       float64(r.Similarity),
			CreatedAtMs: 0,
		})
		if len(hits) >= limit {
			break
		}
	}
	return hits, nil
}

// SearchRecordsFTSForApp kept for callers — delegates to vector search.
func (rm *RecordManager) SearchRecordsFTSForApp(appID, queryText string, tableNames []string, limit int) ([]TenantRAGHit, error) {
	return rm.SearchRecordsVectorForApp(appID, queryText, tableNames, limit)
}

func (rm *RecordManager) mirrorVectorChunkPebble(chunk TenantRAGChunk) {
	rec := map[string]any{
		"chunk_id": chunk.ChunkID, "app_id": chunk.AppID, "source_name": chunk.SourceName,
		"scope_mask": chunk.ScopeMask, "scope_tags": chunk.ScopeTags, "tags": chunk.Tags,
		"created_at_ms": chunk.CreatedAtMs, "summary": chunk.Summary,
		"structure": chunk.Structure, "content": chunk.Content,
	}
	_, _ = rm.CreateRecord(vectorMetaApp, vectorMetaTable, rec, []string{"chunk_id"})
}

func (rm *RecordManager) deleteVectorChunksPebbleBySource(appID, sourceName string) {
	_ = rm.scanTable(vectorMetaApp, vectorMetaTable, func(storageKey string, raw []byte) error {
		var rec map[string]any
		if json.Unmarshal(raw, &rec) != nil {
			return nil
		}
		if fmt.Sprint(rec["app_id"]) != appID || fmt.Sprint(rec["source_name"]) != sourceName {
			return nil
		}
		rm.deleteAtStorageKey(vectorMetaApp, vectorMetaTable, storageKey)
		return nil
	})
}

func scopeMaskForTable(tableName string) int {
	switch tableName {
	case "index":
		return 1
	case "sys_autos":
		return 2
	default:
		return 16
	}
}

// ChunkText splits long text into retrieval-sized chunks (~2200 chars).
func ChunkText(sourceName, text string, maxChunk int) []string {
	text = strings.TrimSpace(text)
	if text == "" {
		return nil
	}
	if maxChunk <= 0 {
		maxChunk = 2200
	}
	if len(text) <= maxChunk {
		return []string{text}
	}
	var chunks []string
	for len(text) > 0 {
		if len(text) <= maxChunk {
			chunks = append(chunks, text)
			break
		}
		cut := maxChunk
		if idx := strings.LastIndex(text[:cut], "\n\n"); idx > maxChunk/2 {
			cut = idx
		}
		chunks = append(chunks, strings.TrimSpace(text[:cut]))
		text = strings.TrimSpace(text[cut:])
	}
	_ = sourceName
	return chunks
}

File: backend-rust/src/data/tenant_rag.rs
//! Tenant RAG vector search — mirrors `backend-go/internal/data/tenant_rag.go`.

use std::collections::HashMap;

use super::vector_store::{VectorQueryHit, VECTOR_COLL_TENANT_RAG};

const DEFAULT_TENANT_RAG_TOP_K: usize = 6;

#[derive(Debug, Clone)]
pub struct TenantRagChunk {
    pub chunk_id: String,
    pub app_id: String,
    pub source_name: String,
    pub scope_mask: i32,
    pub scope_tags: String,
    pub tags: String,
    pub created_at_ms: i64,
    pub summary: String,
    pub structure: String,
    pub content: String,
}

#[derive(Debug, Clone)]
pub struct TenantRagHit {
    pub chunk_id: String,
    pub app_id: String,
    pub source_name: String,
    pub scope_mask: i32,
    pub tags: String,
    pub summary: String,
    pub content: String,
    pub score: f64,
    pub created_at_ms: i64,
}

impl super::RecordManager {
    pub fn ensure_tenant_rag_schema(&self) -> Result<(), String> {
        if self.vector_store.is_none() {
            return Err("vector store unavailable".into());
        }
        Ok(())
    }

    pub fn upsert_tenant_rag_chunk(&self, chunk: TenantRagChunk) -> Result<(), String> {
        self.ensure_tenant_rag_schema()?;
        if chunk.chunk_id.is_empty() || chunk.app_id.is_empty() || chunk.content.trim().is_empty() {
            return Ok(());
        }
        let vs = self.vector_store.as_ref().unwrap();
        let meta = tenant_chunk_meta(&chunk);
        let text = tenant_chunk_embed_text(&chunk);
        vs.upsert_doc(VECTOR_COLL_TENANT_RAG, &chunk.chunk_id, &meta, &text)
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub fn delete_tenant_rag_source(&self, app_id: &str, source_name: &str) {
        if app_id.is_empty() || source_name.is_empty() {
            return;
        }
        if let Some(vs) = &self.vector_store {
            let mut where_clause = HashMap::new();
            where_clause.insert("app_id".into(), app_id.to_string());
            where_clause.insert("source_name".into(), source_name.to_string());
            let _ = vs.delete_where(VECTOR_COLL_TENANT_RAG, &where_clause);
        }
    }

    pub fn search_tenant_rag(
        &self,
        app_id: &str,
        query_text: &str,
        scope_mask: i32,
        limit: usize,
    ) -> Result<Vec<TenantRagHit>, String> {
        let vs = self
            .vector_store
            .as_ref()
            .ok_or_else(|| "vector store unavailable".to_string())?;
        if app_id.is_empty() || query_text.trim().is_empty() {
            return Ok(Vec::new());
        }
        let limit = if limit == 0 {
            DEFAULT_TENANT_RAG_TOP_K
        } else {
            limit
        };
        let mut where_clause = HashMap::new();
        where_clause.insert("app_id".into(), app_id.to_string());
        let results = vs
            .query(VECTOR_COLL_TENANT_RAG, query_text, &where_clause, limit * 3)
            .map_err(|e| e.to_string())?;
        let mut hits = Vec::new();
        for r in results {
            let h = hit_to_tenant_rag(r);
            if !scope_mask_matches(h.scope_mask, scope_mask) {
                continue;
            }
            hits.push(h);
            if hits.len() >= limit {
                break;
            }
        }
        Ok(hits)
    }

    pub fn search_records_vector_for_app(
        &self,
        app_id: &str,
        query_text: &str,
        table_names: &[String],
        limit: usize,
    ) -> Result<Vec<TenantRagHit>, String> {
        let vs = self
            .vector_store
            .as_ref()
            .ok_or_else(|| "vector store unavailable".to_string())?;
        if app_id.is_empty() || query_text.trim().is_empty() {
            return Ok(Vec::new());
        }
        let limit = if limit == 0 {
            DEFAULT_TENANT_RAG_TOP_K
        } else {
            limit
        };
        let default_tables = [
            "csm_roles", "csm_depts", "csm_branches", "index", "sys_autos",
        ];
        let tables: Vec<&str> = if table_names.is_empty() {
            default_tables.to_vec()
        } else {
            table_names.iter().map(String::as_str).collect()
        };
        let table_set: HashMap<&str, ()> = tables.into_iter().map(|t| (t, ())).collect();

        let mut where_clause = HashMap::new();
        where_clause.insert("app_id".into(), app_id.to_string());
        let results = vs
            .query(
                super::vector_store::VECTOR_COLL_RECORDS,
                query_text,
                &where_clause,
                limit * 4,
            )
            .map_err(|e| e.to_string())?;

        let mut hits = Vec::new();
        for r in results {
            let table_name = r.metadata.get("table_name").cloned().unwrap_or_default();
            if !table_set.contains_key(table_name.as_str()) {
                continue;
            }
            let scope = scope_mask_for_table(&table_name);
            hits.push(TenantRagHit {
                chunk_id: r.doc_id,
                app_id: app_id.to_string(),
                source_name: table_name,
                scope_mask: scope,
                tags: String::new(),
                summary: r.metadata.get("title").cloned().unwrap_or_default(),
                content: r.content,
                score: f64::from(r.score.max(0.0)),
                created_at_ms: 0,
            });
            if hits.len() >= limit {
                break;
            }
        }
        Ok(hits)
    }
}

fn tenant_chunk_meta(chunk: &TenantRagChunk) -> HashMap<String, String> {
    let mut m = HashMap::new();
    m.insert("app_id".into(), chunk.app_id.clone());
    m.insert("source_name".into(), chunk.source_name.clone());
    m.insert("scope_mask".into(), chunk.scope_mask.to_string());
    m.insert("scope_tags".into(), chunk.scope_tags.clone());
    m.insert("tags".into(), chunk.tags.clone());
    m.insert("created_at_ms".into(), chunk.created_at_ms.to_string());
    m.insert("summary".into(), chunk.summary.clone());
    m.insert("structure".into(), chunk.structure.clone());
    m
}

fn tenant_chunk_embed_text(chunk: &TenantRagChunk) -> String {
    format!(
        "{}\n{}\n{}",
        chunk.summary, chunk.structure, chunk.content
    )
}

fn hit_to_tenant_rag(r: VectorQueryHit) -> TenantRagHit {
    let scope_mask = r
        .metadata
        .get("scope_mask")
        .and_then(|s| s.parse().ok())
        .unwrap_or(0);
    let created_at_ms = r
        .metadata
        .get("created_at_ms")
        .and_then(|s| s.parse().ok())
        .unwrap_or(0);
    let score = f64::from(r.score.max(0.0));
    TenantRagHit {
        chunk_id: r.doc_id,
        app_id: r.metadata.get("app_id").cloned().unwrap_or_default(),
        source_name: r.metadata.get("source_name").cloned().unwrap_or_default(),
        scope_mask,
        tags: r.metadata.get("tags").cloned().unwrap_or_default(),
        summary: r.metadata.get("summary").cloned().unwrap_or_default(),
        content: r.content,
        score,
        created_at_ms,
    }
}

fn scope_mask_matches(doc_scope_mask: i32, filter_mask: i32) -> bool {
    if filter_mask == 0 {
        return true;
    }
    (doc_scope_mask & filter_mask) != 0
}

fn scope_mask_for_table(table_name: &str) -> i32 {
    match table_name {
        "index" => 1,
        "sys_autos" => 2,
        _ => 16,
    }
}

pub fn chunk_text(source_name: &str, text: &str, max_chunk: usize) -> Vec<String> {
    let text = text.trim();
    if text.is_empty() {
        return Vec::new();
    }
    let max_chunk = if max_chunk == 0 { 2200 } else { max_chunk };
    if text.len() <= max_chunk {
        return vec![text.to_string()];
    }
    let mut chunks = Vec::new();
    let mut rest = text;
    while !rest.is_empty() {
        if rest.len() <= max_chunk {
            chunks.push(rest.to_string());
            break;
        }
        let mut cut = max_chunk;
        if let Some(idx) = rest[..cut].rfind("\n\n") {
            if idx > max_chunk / 2 {
                cut = idx;
            }
        }
        chunks.push(rest[..cut].trim().to_string());
        rest = rest[cut..].trim();
    }
    let _ = source_name;
    chunks
}
